File: grabmem.go
// a program to grab memory over time

package main
import (
 "fmt"
 "io"
 //"bytes"
 "flag"
 "time"
 "crypto/rand"
)
var  memsize, step, max int
var interval, now, then time.Duration
var myMemory = make([]byte, step)
//var myMemory  bytes.Buffer

func usage() {
    fmt.Println("a program to suck up memory")
}

func init() {
    flag.IntVar(&memsize, "step",1, "step size for memory increases")
    flag.IntVar(&max, "max",5, "maximum memory to allcate")
    flag.DurationVar(&interval, "interval", 1, "time between memory steps")
    flag.Parse()
}

func fill() {
    //myMemory.Grow(step)
    bytesWritten, err := io.ReadFull(rand.Reader, myMemory)
    if bytesWritten != len(myMemory) || err != nil {
        fmt.Println("error:",err)
        return
    }
    time.Sleep(interval)
}

func main () {
    myMemory = make([]byte, step)
    for memsize <= max {
      myMemory = append(myMemory , [step]byte  )
      fill()
    }
    usage()
}

File: grabmem.rs
//we'll figure out cli parsing later but its 'extern crate getopts;
extern crate getopts;
use getopts::Options;
use std::{env, time};
fn usage( ){
    println!("usage: grabmem max rate step");
    println!("this program allocates memory in RATE second intervals, increasing by STEP megabytes of memory, untill memory reaches MAX.\n");
}
fn grab(size: i64) {
    myMemory = size; //will this get optimized away if we never use it?
    memsize += size;
    println!("holding {} memory", memsize);
}


fn main(){  
    let arg: Vec<String> = env::args().collect();

    usage();
    const max  = &arg[1];
    const rate = &arg[2];
    const step = &arg[3].parse::<i64>();
    let duration = time::Duration::from_secs(rate);
    let mut memsize = 0i64 ;
    let mut myMemory = [i64;step] ;

    loop { 
        if memsize >= max {
            println!("hit max memory");
            break;
        }
        grab(step);
        thread::sleep(duration);
    }
}

